src/settings/card_health_check.c

```diff
@@ -0,0 +1,206 @@
+#include "card_operations.h"
+#include "constant_texts.h"
+#include "core_error.h"
+#include "flash_api.h"
+#include "settings_api.h"
+#include "ui_core_confirm.h"
+#include "ui_screens.h"
+#include "ui_state_machine.h"
+
+/*****************************************************************************
+ * EXTERN VARIABLES
+ *****************************************************************************/
+
+/*****************************************************************************
+ * PRIVATE MACROS AND DEFINES
+ *****************************************************************************/
+
+/*****************************************************************************
+ * PRIVATE TYPEDEFS
+ *****************************************************************************/
+
+/*****************************************************************************
+ * STATIC FUNCTION PROTOTYPES
+ *****************************************************************************/
+
+/*****************************************************************************
+ * STATIC VARIABLES
+ *****************************************************************************/
+
+/*****************************************************************************
+ * GLOBAL VARIABLES
+ *****************************************************************************/
+
+/*****************************************************************************
+ * STATIC FUNCTIONS
+ *****************************************************************************/
+
+/*****************************************************************************
+ * GLOBAL FUNCTIONS
+ *****************************************************************************/
+void card_health_check(void) {
+  wallet_list_t wallets_in_card = {0};
+
+  card_fetch_wallet_list_config_t configuration = {0};
+  configuration.operation.acceptable_cards = ACCEPTABLE_CARDS_ALL;
+  configuration.operation.skip_card_removal = true;
+  configuration.operation.expected_family_id = get_family_id();
+  configuration.frontend.heading = NULL;
+  configuration.frontend.msg = ui_text_card_health_check_start;
+
+  card_fetch_wallet_list_response_t response = {0};
+  response.wallet_list = &wallets_in_card;
+  response.card_info.tapped_card = 0;
+  response.card_info.recovery_mode = 0;
+  response.card_info.status = 0;
+  response.card_info.tapped_family_id = NULL;
+
+  // P0 abort is the only condition we want to exit the flow
+  // Card abort error will be explicitly shown here as error codes
+  card_error_type_e status = card_fetch_wallet_list(&configuration, &response);
+  if (CARD_OPERATION_P0_OCCURED == status) {
+    return;
+  }
+
+  // If the tapped card is not paired, it is a terminal case in the flow
+  // Card operation will return ABORT but the status will signify SW_NO_ERROR
+  if (CARD_OPERATION_ABORT_OPERATION == status &&
+      SW_NO_ERROR == response.card_info.status) {
+    return;
+  }
```

Comment:
Maybe we should have a flag in the `NFC_connection_data` structure, that tells if card was paired or not

Reply:
I have added this based on discussion. Please re-review